### ISSUE TITLE
BufferedSource.peek()

### ISSUE THREAD

**Issue by swankjesse** [square/okio#506]
We’ve discussed an idea for a fancy peek API. Here’s an API sketch:

```
interface BufferedSource {
  /**
   * Returns a new `BufferedSource` that can read data from this `BufferedSource`
   * without consuming it. The returned source becomes invalid once this source
   * is next read or closed.
   *
   * For example, we can use `peek()` to lookahead and read the same data multiple
   * times.
   * 
   * ```
   * val buffer = Buffer()
   * buffer.writeUtf8("abcdefghij")
   *
   * buffer.readUtf8(3) // returns "abc", buffer contains "defghi"
   *
   * val peek = buffer.peek()
   * peek.readUtf8(3) // returns "def", buffer contains "defghi"
   * peek.readUtf8(3) // returns "ghi", buffer contains "defghi"
   *
   * buffer.readUtf8() // returns "def", buffer contains "ghi"
   * ```
   */
  BufferedSource peek()
}
```

Behind the scenes the peeking source would copy data from the upstream buffer into its own private buffer.

**Comment by swankjesse** [comment]
Found the original mention:
https://github.com/square/okio/issues/362#issuecomment-374725968

**Comment by JakeWharton** [comment]
And most importantly: fills the upstream buffer when you exceed it's currently buffered amount.

**Comment by JakeWharton** [comment]
Most importantly for the original use case on the linked issue, that is.

**Comment by swankjesse** [comment]
Yes! It can outrun upstream, which causes more bytes to be buffered into memory.

Which reminds me that I would like to someday get back to LruCache2 and Relay.java. It attempts to solve the same problem but it uses a file to prevent unbounded buffering!

**Comment by bnorm** [comment]
I have a working prototype for this but I have a question about: "The returned source becomes invalid once this source is next read or closed." Is the desire for the peek source to throw an exception after the base source is read from? Or is this just something the user needs to be aware of when using the peek source?

**Comment by swankjesse** [comment]
I think the behavior we want will be too inefficient to build if the peeking source keeps working as its upstream is used.

So instead we have a hard rule “don’t use the peeking source after you’ve used its upstream source”. We can enforce that with unchecked exceptions or not, depending on what the cost of enforcement is.

**Comment by bnorm** [comment]
What I've found is that enforcing no use of the peek source after a read from the upstream source is probably going to be expensive and complicated. I found it would require every method in BufferedSource to mark any peeking sources as unusable.

My simple implementation was easy enough, but assumes the user understands when to stop using the peek source.
```
override fun peek(): BufferedSource {
  return object : Source {
    var pos = 0L

    override fun read(sink: Buffer, byteCount: Long): Long {
      request(pos + byteCount)
      val toCopy = minOf(byteCount, buffer.size - pos)
      if (toCopy <= 0L) { // protect from the buffer being consumed
        return -1L
      }

      buffer.copyTo(sink, pos, toCopy)
      pos += toCopy
      return toCopy
    }

    override fun timeout(): Timeout {
      return source.timeout()
    }

    override fun close() {
      // Buffered source takes care of this
    }
  }.buffer()
}
```

**Comment by swankjesse** [comment]
@bnorm that’s brilliant. Send a PR?

I’ve got a wacky idea on how to detect upstream changes. It’s inspired by HashMap.[modCount](https://android.googlesource.com/platform/libcore.git/+/android-3.2.4_r1/luni/src/main/java/java/util/HashMap.java#110).

Here’s the gist of it:

```
override fun peek(): BufferedSource {
  return object : Source {
    var expectedUpstreamFirstSegment = ...
    var expectedUpstreamOffset = ...
    ...

    override fun read(sink: Buffer, byteCount: Long): Long {
      require(expectedUpstreamFirstSegment == buffer.segment)
      require(expectedUpstreamOffset == buffer.segment.offset)
      ...
    }

    ...
  }
}
```

It’s probably a little annoying to implement because the first segment will initially be null, and growing the buffer can change the first segment. But the peek could potentially do this bookkeeping.

```

**Comment by bnorm** [comment]
I actually tried this but it get's a bit more complicated because we want to return a BufferedSource. The peek source will try and grab a full segment on the first read, which means secondary small reads from the peek, as long as they are still in the buffer, won't need to call the underlying Source.read.

My thought was that I want it to either fail right away, or not fail at all and have no guarantees on what data is returned. I'm not sure how I feel about it failing after a few unsafe reads of the peek source.

**Comment by swankjesse** [comment]
Gotcha. It won't catch all programming mistakes. But it should still prevent data corruption because we'll hault before we return unexpected data.
